kubeflow/trainer/rhai/instrumentation/speculator_progression.py

```
@@ -0,0 +1,347 @@
+                        elif msg["epoch"] >= 1 and _max_step_in_epoch0 >= 0:
+                            _steps_per_epoch = _max_step_in_epoch0 + 1
+            except (KeyError, TypeError, ValueError) as e:
+                print(f"[Kubeflow] Warning: Failed to parse metrics record: {e}", flush=True)
+
+    class SpeculatorMetricsHTTPHandler(http.server.BaseHTTPRequestHandler):
+        """HTTP handler that serves mode-aware progress to the controller."""
+
+        def do_GET(self):
+            try:
+                transformed = self._get_progress()
+            except Exception as e:
```

Ruff (0.16.1)

[warning] Do not catch blind exception: `Exception`

(BLE001)